### PR TITLE
[JUJU-4012] Wire up watcher registry

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following wires up the watcher registry to the pinger timeout watcher.
This is the first watcher to take advantage of the watcher registry. We're
deprecating the `Resource` type in favour of a real `worker.Worker`, that
will correctly monitor the lifecycle of a watcher.

This supersedes https://github.com/juju/juju/pull/15767 which required too
much churn. Instead, it will require every person to migrate each watcher
over to the `WatcherRegistry` as we move along. This is the cleanest 
bulletproof way to ensure that we don't break anything as we go.

As a driveby, I cleaned up the API server to make it readable.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Ensure there are no errors when adding and removing models.

```sh
$ juju bootstrap lxd test --build-agent
$ juju debug-log -m controller
$ juju add-model default
$ juju destroy-model default
```

### CODE REVIEW

File: apiserver/admin.go [hunk 1]
```diff
@@ -533,7 +533,14 @@ func (a *admin) maintenanceInProgress() bool {
 	return !a.srv.upgradeComplete()
 }
 
-func setupPingTimeoutDisconnect(clock clock.Clock, root *apiHandler, entity state.Entity) error {
+// PingRootHandler is the interface that the root handler must implement
+// to allow the pinger to be registered.
+type PingRootHandler interface {
```
Comment (manadart) [inline apiserver/admin.go:538]: Any reason to export this? It isn't used elsewhere.

File: apiserver/facade/watcherregistry.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package facade
+
+import "github.com/juju/worker/v3"
+
+// WatcherRegistry holds all the watchers for a connection.
+// It allows the registration of watchers that will be cleaned up when a
+// connection terminates.
+type WatcherRegistry interface {
```
Comment (manadart) [inline apiserver/facade/watcherregistry.go:11]: Technically it's a worker registry right? Watcher being the specific.

File: apiserver/root.go [hunk 3]
```diff
@@ -143,16 +159,125 @@ func (r *apiHandler) Authorizer() facade.Authorizer {
 	return r
 }
 
-func (r *apiHandler) getRpcConn() *rpc.Conn {
-	return r.rpcConn
+// CloseConn closes the underlying connection.
+func (r *apiHandler) CloseConn() error {
```
Comment (manadart) [inline apiserver/root.go:163]: Nice touch.

File: apiserver/root.go [hunk 4]
```diff
@@ -143,16 +159,125 @@ func (r *apiHandler) Authorizer() facade.Authorizer {
...
+// AuthMachineAgent returns whether the current client is a machine agent.
+// TODO(controlleragent) - add AuthControllerAgent function
+func (r *apiHandler) AuthMachineAgent() bool {
+	_, isMachine := r.GetAuthTag().(names.MachineTag)
+	_, isControllerAgent := r.GetAuthTag().(names.ControllerAgentTag)
+	return isMachine || isControllerAgent
+}
+
+// AuthModelAgent return whether the current client is a model agent
+func (r *apiHandler) AuthModelAgent() bool {
```
Comment (manadart) [inline apiserver/root.go:186]: Seems to me we could something nicer than these with generics.
Not the concern of this patch of course.

**Review by manadart — APPROVED** [review]
Nice changes.

File: apiserver/facade/watcherregistry.go (hunk 2 above)
Comment (SimonRichardson) [reply apiserver/facade/watcherregistry.go:11]: I wanted to be specific so people feel bad about shoving random workers into the registry. For example, the `pingTimeout` and `crossModelOfferURL` conform to a worker, but aren't actually workers. They don't have a lifecycle in the true sense, they're just inturned into a bucket and expected to be around at some point.

File: apiserver/admin.go (hunk 1 above)
Comment (SimonRichardson) [reply apiserver/admin.go:538]: I was thinking there should be a test for this.

File: worker/dbaccessor/tracker_test.go [hunk 5]
```diff
@@ -295,7 +295,7 @@ func (s *trackedDBWorkerSuite) TestWorkerAttemptsToVerifyDBButSucceedsWithDiffer
 	// In-theatre this will be OK, because a DB in an error state recoverable
 	// by reconnecting will be replaced within the default retry strategy's
 	// backoff/repeat loop.
-	timeout := time.After(testing.ShortWait)
```
Comment (SimonRichardson) [inline worker/dbaccessor/tracker_test.go:298]: This isn't enough if you have race and have `-count=100` :scream:

**Pull request merged**